### PR TITLE
Feature/init actor state tree

### PR DESCRIPTION
**Pull request by turuslan**
### Description of the Change
Add [state tree](https://github.com/filecoin-project/lotus/blob/master/chain/state/statetree.go).
Add [init actor](https://github.com/filecoin-project/lotus/blob/master/chain/actors/actor_init.go) except vm methods.
Extend IpfsDatastore with convenient methods combinig cbor encoding and decoding: getCbor, setCbor.
Extend Hamt with convenient methods combinig cbor encoding and decoding: getCbor, setCbor.

### Benefits
State tree and init actor state.

### Possible Drawbacks 
None

### CODE REVIEW

File: core/primitives/address/address.hpp [hunk 1]
```diff
@@ -65,6 +65,8 @@ namespace fc::primitives::address {
      */
     bool isKeyType() const;
 
+    static Address makeFromId(uint64_t id);
```
Comment (Alexey-N-Chernyshov) [inline core/primitives/address/address.hpp]: What is an id? Please, leave comment.

File: core/storage/ipfs/datastore.hpp [hunk 2]
```diff
@@ -51,8 +53,39 @@ namespace fc::storage::ipfs {
      * @return success if removed or didn't exist, error otherwise
      */
     virtual outcome::result<void> remove(const CID &key) = 0;
-  };
 
+    /** Compute CID from bytes */
+    inline static outcome::result<CID> cid(gsl::span<const uint8_t> bytes) {
```
Comment (Alexey-N-Chernyshov) [inline core/storage/ipfs/datastore.hpp]: I don't like an idea to put cid converters into datastore interface. Datastore is a simple and clean interface representing underlying storage. With cid converters it turns into datastore-with-cid-converters interface. Shortly speaking, please move cid-related method to CID or separate interface and keep interfaces clean.

File: core/primitives/address/address.hpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/primitives/address/address.hpp]: @param id - number assigned to actors in a Filecoin Chain

File: core/primitives/address/address.cpp [hunk 3]
```diff
@@ -43,6 +43,11 @@ namespace fc::primitives::address {
         [](const BLSPublicKeyHash &v) { return Protocol::BLS; });
   }
 
+  Address Address::makeFromId(uint64_t id) {
+    // TODO(turuslan): FIL-118 remove hardcoded TESTNET
```
Comment (Alexey-N-Chernyshov) [inline core/primitives/address/address.cpp]: ❤️

File: core/storage/ipfs/datastore.hpp (hunk 2 above)
Comment (turuslan) [reply core/storage/ipfs/datastore.hpp]: No. Go implementation provides such interface.
Template type T is like go "interface{}".

File: core/storage/ipfs/datastore.hpp [hunk 4]
```diff
@@ -51,8 +53,39 @@ namespace fc::storage::ipfs {
      * @return success if removed or didn't exist, error otherwise
      */
     virtual outcome::result<void> remove(const CID &key) = 0;
-  };
 
+    /** Compute CID from bytes */
```
Comment (igor-egorov) [inline core/storage/ipfs/datastore.hpp]: Triple slash can be used for one-line docstrings
```suggestion
    /// Compute CID from bytes
```

File: .clang-format [hunk 5]
```diff
@@ -7,3 +7,4 @@ PointerAlignment: Right
 BinPackArguments: false
 BinPackParameters: false
 AllowShortFunctionsOnASingleLine: Empty
+IncludeBlocks: Preserve
```
Comment (masterjedy) [inline .clang-format]: great

File: core/vm/actor/actor.hpp [hunk 6]
```diff
@@ -8,28 +8,34 @@
...
-  using primitives::BigInt;
   using libp2p::multi::ContentIdentifier;
+  using primitives::address::Address;
+  using primitives::BigInt;
 
   /** Common actor state interface */
   struct Actor {
-    ContentIdentifier code;
-    ContentIdentifier head;
+    ContentIdentifier code{{}, {}, libp2p::multi::Multihash::create({}, {}).value()};
```
Comment (masterjedy) [inline core/vm/actor/actor.hpp]: consider adding default constructor to ContentIdentifier or a function creating empty ContentIdentifier.

File: core/vm/actor/init_actor.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  using storage::ipfs::IpfsDatastore;
+
+  /// Init actor state
+  struct InitActorState {
+    /// Allocate new id address
+    outcome::result<Address> addActor(std::shared_ptr<IpfsDatastore> store,
+                                      const Address &address);
+
+    ContentIdentifier address_map{
+        {}, {}, libp2p::multi::Multihash::create({}, {}).value()};
```
Comment (masterjedy) [inline core/vm/actor/init_actor.hpp]: same

File: core/storage/ipfs/datastore.hpp [hunk 8]
```diff
@@ -51,8 +53,39 @@ namespace fc::storage::ipfs {
...
+      return cid(bytes);
+    }
+
+    /// Store CBOR encoded value
+    template <typename T>
+    outcome::result<CID> setCbor(const T &value) {
+      OUTCOME_TRY(bytes, codec::cbor::encode(value));
+      OUTCOME_TRY(cid, cid(bytes));
+      OUTCOME_TRY(set(cid, Value(bytes)));
+      return std::move(cid);
```
Comment (masterjedy) [inline core/storage/ipfs/datastore.hpp]: is clang-tidy ok with it?

File: test/core/vm/actor/CMakeLists.txt [hunk 9]
```diff
@@ -11,3 +11,13 @@ target_link_libraries(actor_test
     cbor
     hexutil
     )
+
+addtest(init_actor_test
+    init_actor_test.cpp
+    )
+target_link_libraries(init_actor_test
+    actor
```
Comment (masterjedy) [inline test/core/vm/actor/CMakeLists.txt]: test's target_link_libraries section should be as small as possible, ideally containing one library - tested one and probably testutils. if it is not compiled you probably should link them to the original library.

File: test/core/vm/actor/actor_test.cpp [hunk 10]
```diff
@@ -36,10 +37,9 @@ TEST(ActorTest, CidGoCompatibility) {
 /** Actor CBOR encoding and decoding */
 TEST(ActorTest, ActorGoCompatibility) {
   auto go = "84d82a52000155000d66696c2f312f6163636f756e74d82a4f000155000a66696c2f312f63726f6e03420005"_unhex;
-  fc::vm::actor::Actor actor{fc::vm::actor::kEmptyObjectCid, fc::vm::actor::kEmptyObjectCid, 0, 0};
 
-  fc::codec::cbor::CborDecodeStream(go) >> actor;
-  EXPECT_EQ((fc::codec::cbor::CborEncodeStream() << actor).data(), go);
+  EXPECT_OUTCOME_TRUE(actor, fc::codec::cbor::decode<fc::vm::actor::Actor>(go));
```
Comment (masterjedy) [inline test/core/vm/actor/actor_test.cpp]: sometimes `using ...` makes life easier.

File: test/core/vm/actor/init_actor_test.cpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <gtest/gtest.h>
+#include "primitives/address/address_codec.hpp"
+#include "storage/hamt/hamt.hpp"
+#include "storage/ipfs/impl/in_memory_datastore.hpp"
+#include "testutil/cbor.hpp"
+
+using fc::vm::actor::InitActorState;
+
+/** Init actor state CBOR */
```
Comment (masterjedy) [inline test/core/vm/actor/init_actor_test.cpp]: according to our practices, tests should be supplied with /** @given ... @when ... @then ... */ description.

File: test/core/vm/actor/init_actor_test.cpp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+TEST(InitActorTest, ActorState) {
```
Comment (masterjedy) [inline test/core/vm/actor/init_actor_test.cpp]: According to our practices, test names should have `Success` or `Fail` suffixes when applicable.
E.g. `InitActorStateSuccess` or `AddActorSuccess`.

File: test/core/vm/state/state_tree_test.cpp [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+class StateTreeTest : public ::testing::Test {
+ public:
+  std::shared_ptr<fc::storage::ipfs::IpfsDatastore> store_{
+      std::make_shared<fc::storage::ipfs::InMemoryDatastore>()};
+  StateTree tree_{store_};
+};
+
+/**
+ * @given State tree and actor state
+ * @when set
```
Comment (masterjedy) [inline test/core/vm/state/state_tree_test.cpp]: pls make tests descriptions more detailed, 
e.g. 
@given State tree, actor state and predefined (key, value) pair
@when set key, value pair in the tree  
...
maybe my example is not good enough, just to show.

File: test/testutil/cbor.hpp [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+#ifndef CPP_FILECOIN_TEST_TESTUTIL_CBOR_HPP
+#define CPP_FILECOIN_TEST_TESTUTIL_CBOR_HPP
+
+#include "codec/cbor/cbor.hpp"
+#include "testutil/literals.hpp"
+#include "testutil/outcome.hpp"
+
+template <typename T>
```
Comment (masterjedy) [inline test/testutil/cbor.hpp]: pls add description to the function

File: test/core/vm/actor/init_actor_test.cpp (hunk 11 above)
Comment (turuslan) [reply test/core/vm/actor/init_actor_test.cpp]: Lotus-compatible CBOR encoding tests are common and simple.
Big description would be excessive.
Most of them would just use `expectEncodeAndReencode`.

File: test/core/vm/actor/init_actor_test.cpp (hunk 12 above)
Comment (turuslan) [reply test/core/vm/actor/init_actor_test.cpp]: Maybe we can add `Cbor` suffixes for lotus-compatible CBOR encoding tests.

File: core/storage/hamt/hamt.hpp [hunk 15]
```diff
@@ -113,10 +111,22 @@ namespace fc::storage::hamt {
     /** Remove value by key, does not write to storage */
     outcome::result<void> remove(const std::string &key);
     /** Write changes made by set and remove to storage */
-    outcome::result<void> flush();
+    outcome::result<CID> flush();
     /** Apply visitor for key value pairs */
     outcome::result<void> visit(const Visitor &visitor);
 
+    template <typename T>
```
Comment (masterjedy) [inline core/storage/hamt/hamt.hpp]: add descriptions

File: core/storage/ipfs/datastore.hpp [hunk 16]
```diff
@@ -51,8 +53,39 @@ namespace fc::storage::ipfs {
...
+    outcome::result<CID> setCbor(const T &value) {
+      OUTCOME_TRY(bytes, codec::cbor::encode(value));
+      OUTCOME_TRY(key, cid(bytes));
+      OUTCOME_TRY(set(key, Value(bytes)));
+      return std::move(key);
+    }
+
+    /// Get CBOR encoded value by CID
+    template <typename T>
+    outcome::result<T> getCbor(const CID &key) {
```
Comment (Alexey-N-Chernyshov) [inline core/storage/ipfs/datastore.hpp]:
```suggestion
    outcome::result<T> getCbor(const CID &key) const {
```

File: core/storage/ipfs/datastore.hpp [hunk 17]
```diff
@@ -51,8 +53,39 @@ namespace fc::storage::ipfs {
...
+    /// Store CBOR encoded value
+    template <typename T>
+    outcome::result<CID> setCbor(const T &value) {
+      OUTCOME_TRY(bytes, codec::cbor::encode(value));
+      OUTCOME_TRY(key, cid(bytes));
+      OUTCOME_TRY(set(key, Value(bytes)));
+      return std::move(key);
+    }
+
+    /// Get CBOR encoded value by CID
```
Comment (Alexey-N-Chernyshov) [inline core/storage/ipfs/datastore.hpp]: But method returns decoded value, isn't it?

File: core/storage/hamt/hamt.hpp [hunk 18]
```diff
@@ -125,7 +137,7 @@ namespace fc::storage::hamt {
     outcome::result<void> remove(Node &node,
                                  gsl::span<const size_t> indices,
                                  const std::string &key);
-    outcome::result<void> cleanShard(Node::Item &item);
+    static outcome::result<void> cleanShard(Node::Item &item);
     outcome::result<void> flush(Node::Item &item);
     outcome::result<void> loadItem(Node::Item &item);
```
Comment (Alexey-N-Chernyshov) [inline core/storage/hamt/hamt.hpp]: Can be marked 'const'
```suggestion
    outcome::result<void> loadItem(Node::Item &item) const;
```

**Pull request merged**